File: lib/models/user model/user_model.dart
import 'package:json_annotation/json_annotation.dart';

part 'user_model.g.dart';

@JsonSerializable()
class Users {
  final String id;

  final String email, name;
  final String? password;
  final DateTime? dob;
  final bool emailVerified;
  final String? profileImageURL;
  final String provider;
  final List<String> workExperience;
  final List<String> education;
  final List<String> skills;

  final List<WorkExperience> workExperience_data;
  final List<Education> education_data;
  final List<Skills> skills_data;

  Users({
    this.workExperience = const [],
    this.education = const [],
    this.skills = const [],
    required this.id,
    required this.email,
    required this.name,
    required this.password,
    required this.dob,
    required this.emailVerified,
    this.profileImageURL,
    required this.provider,
    this.workExperience_data = const [],
    this.education_data = const [],
    this.skills_data = const [],
  });

  factory Users.fromJson(Map<String, dynamic> json) => _$UsersFromJson(json);

  Map<String, dynamic> toJson() => _$UsersToJson(this);
  @override
  bool operator == (covariant Users other) {
    return other.id == id;
  }





}

@JsonSerializable()
class WorkExperience {
  final String id;

  final String position, company, jobType;
  final String? location;
  final DateTime start;
  final DateTime? end;

  WorkExperience({
    required this.id,
    required this.position,
    required this.company,
    required this.jobType,
    this.location,
    required this.start,
    required this.end,
  });

  factory WorkExperience.fromJson(Map<String, dynamic> json) =>
      _$WorkExperienceFromJson(json);

  Map<String, dynamic> toJson() => _$WorkExperienceToJson(this);
}

@JsonSerializable()
class Education {
  final String id;

  final String institution, degree, field;
  final String? location;
  final DateTime start;
  final DateTime? end;

  Education({
    required this.id,
    required this.institution,
    required this.degree,
    required this.field,
    this.location,
    required this.start,
    required this.end,
  });

  factory Education.fromJson(Map<String, dynamic> json) =>
      _$EducationFromJson(json);

  Map<String, dynamic> toJson() => _$EducationToJson(this);
}

@JsonSerializable()
class Skills {
  final String id;

  final String skill, expertise;

  Skills({
    required this.id,
    required this.skill,
    required this.expertise,
  });

  factory Skills.fromJson(Map<String, dynamic> json) => _$SkillsFromJson(json);

  Map<String, dynamic> toJson() => _$SkillsToJson(this);
}

File: server/models/User Profile Models/user.model.ts
import { Model, Schema, model, } from "mongoose";
import * as bcrypt from "bcrypt";
import { IWorkExperience, WorkExperienceSchema } from "./sub_documents/work_experience.model";
import { EducationSchema, IEducation } from "./sub_documents/education.model";
import { ISkills, SkillsSchema } from "./sub_documents/skills.model";

export interface IUser {
    email: string;
    password: string;
    name: string,
    dob: Date,
    emailVerified: boolean,
    profileImageURL: string,
    provider: string,
    workExperience?: Array<IWorkExperience>,
    education?: Array<IEducation>,
    skills?: Array<ISkills>,
}

const UserSchema = new Schema<IUser, Model<IUser>>({
    emailVerified: Boolean,
    name: {
        type: String,
        required: true,
    },
    email: {
        type: String,
        required: true,
    },
    password: String,
    dob: Date,
    profileImageURL: String,
    provider: String,
    workExperience: [WorkExperienceSchema],
    education: [EducationSchema],
    skills: [SkillsSchema],

}, { timestamps: true });

export const User = model<IUser, Model<IUser>>('users', UserSchema);